### PR TITLE
Error message when key/nonkey configuration doesn't match actual behaviour

### PR DESCRIPTION
**Pull request by nickbroon**
Add an error message when key/nonkey configuration doesn't match actual behaviour (as reported in #112).

Using a 'traditional' flow definition like that below will result in:

vermont[4729]: Field bgpDestinationAsNumber configured as nonFlowKey will not be Aggregated
vermont[4729]: Field bgpSourceAsNumber configured as nonFlowKey will not be Aggregated
vermont[4729]: Field destinationIPv4PrefixLength configured as nonFlowKey will not be Aggregated
vermont[4729]: Field egressInterface configured as nonFlowKey will not be Aggregated
vermont[4729]: Field ipNextHopIPv4Address configured as nonFlowKey will not be Aggregated
vermont[4729]: Field sourceIPv4PrefixLength configured as nonFlowKey will not be Aggregated

<rule>
        <flowKey>
               <ieName>destinationIPv4Address</ieName>
               <autoAddV4PrefixLength>false</autoAddV4PrefixLength>
        </flowKey>
        <flowKey>
               <ieName>destinationTransportPort</ieName>
        </flowKey>
        <flowKey>
               <ieName>ingressInterface</ieName>
        </flowKey>
        <flowKey>
               <ieName>ipClassOfService</ieName>
        </flowKey>
        <flowKey>
               <ieName>protocolIdentifier</ieName>
        </flowKey>
        <flowKey>
               <ieName>sourceIPv4Address</ieName>
               <autoAddV4PrefixLength>false</autoAddV4PrefixLength>
        </flowKey>
        <flowKey>
               <ieName>sourceTransportPort</ieName>
        </flowKey>
        <nonFlowKey>
               <ieName>bgpDestinationAsNumber</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>bgpSourceAsNumber</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>destinationIPv4PrefixLength</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>egressInterface</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>flowEndMilliseconds</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>flowStartMilliseconds</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>ipNextHopIPv4Address</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>octetDeltaCount</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>packetDeltaCount</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>sourceIPv4PrefixLength</ieName>
        </nonFlowKey>
        <nonFlowKey>
               <ieName>tcpControlBits</ieName>
        </nonFlowKey>
</rule>

### CODE REVIEW

**Comment by nickbroon** [comment]
I'll hopefully have a followup Pull Request that enhances the aggregator so it respects the configuration (by having the default aggregation behaviour for non-keys be 'first flow', as is commonly done on pretty much all other netflow probes)

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp [hunk 1]
```diff
@@ -46,7 +47,7 @@ AggregatorBaseCfg::AggregatorBaseCfg(XMLElement* elem)
 				} else {
 					msg(LOG_CRIT, "Too many rules: %ul\n", MAX_RULES);
 				}
-
```
Comment (ogasser) [inline src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: Unnecessary whitespace change.

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp [hunk 2]
```diff
@@ -140,6 +141,9 @@ Rule::Field* AggregatorBaseCfg::readNonFlowKeyRule(XMLElement* e)
 	ruleField->semantic = ie.getSemantic();
 	ruleField->fieldIe = ie.getFieldIe();
 
+	if (!BaseHashtable::isToBeAggregated(ruleField->type))
+		msg(LOG_ERR, "Field %s configured as nonFlowKey will not be Aggregated", ie.getIeName().c_str());
```
Comment (ogasser) [inline src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: Aggregated -> aggregated

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp [hunk 3]
```diff
@@ -178,6 +182,9 @@ Rule::Field* AggregatorBaseCfg::readFlowKeyRule(XMLElement* e) {
 		ruleField->type.enterprise = ie.getEnterpriseNumber();
 		ruleField->type.length = ie.getIeLength();
 
+		if (BaseHashtable::isToBeAggregated(ruleField->type))
+			msg(LOG_ERR, "Field %s configured as FlowKey will be Aggregated", ie.getIeName().c_str());
```
Comment (ogasser) [inline src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: Aggregated -> aggregated

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp [hunk 4]
```diff
@@ -178,6 +182,9 @@ Rule::Field* AggregatorBaseCfg::readFlowKeyRule(XMLElement* e) {
 		ruleField->type.enterprise = ie.getEnterpriseNumber();
 		ruleField->type.length = ie.getIeLength();
 
+		if (BaseHashtable::isToBeAggregated(ruleField->type))
```
Comment (ogasser) [inline src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: Can you add braces around the then-part of the `if` statement?

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp [hunk 5]
```diff
@@ -140,6 +141,9 @@ Rule::Field* AggregatorBaseCfg::readNonFlowKeyRule(XMLElement* e)
 	ruleField->semantic = ie.getSemantic();
 	ruleField->fieldIe = ie.getFieldIe();
 
+	if (!BaseHashtable::isToBeAggregated(ruleField->type))
```
Comment (ogasser) [inline src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: Can you add braces around the then-part of the `if` statement?

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp (hunk 1 above)
Comment (nickbroon) [reply src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: Doh, rogue editor auto-fixing. Removed.

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp (hunk 2 above)
Comment (nickbroon) [reply src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: Fixed.

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp (hunk 3 above)
Comment (nickbroon) [reply src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: Fixed.

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp (hunk 4 above)
Comment (nickbroon) [reply src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: I'm used to kernel style. :-) Added the braces.

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp (hunk 5 above)
Comment (nickbroon) [reply src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: I'm used to kernel style. :-) Added the braces.

**Comment by ogasser** [comment]
Thanks for the PR and the changes!
